File: Frontend-admin/src/api/ingredientes.ts
import api from './client';
import type { Ingrediente, IngredienteCreate, IngredienteUpdate } from '../types';

interface GetIngredientesParams {
  q?: string;
  es_alergeno?: boolean;
}

export async function getIngredientes(params?: GetIngredientesParams): Promise<Ingrediente[]> {
  const query: Record<string, string> = {};
  if (params?.q) query.q = params.q;
  if (params?.es_alergeno !== undefined) query.es_alergeno = String(params.es_alergeno);
  const res = await api.get<Ingrediente[]>('/ingredientes', { params: query });
  return res.data;
}

export async function getIngredienteById(id: number): Promise<Ingrediente> {
  const res = await api.get<Ingrediente>(`/ingredientes/${id}`);
  return res.data;
}

export async function createIngrediente(data: IngredienteCreate): Promise<Ingrediente> {
  const res = await api.post<Ingrediente>('/ingredientes', data);
  return res.data;
}

export async function updateIngrediente(id: number, data: IngredienteUpdate): Promise<Ingrediente> {
  const res = await api.patch<Ingrediente>(`/ingredientes/${id}`, data);
  return res.data;
}

export async function deleteIngrediente(id: number): Promise<void> {
  await api.delete(`/ingredientes/${id}`);
}

File: Backend/app/routers/ingredientes.py
from typing import List, Optional
from fastapi import APIRouter, Depends, Query, status
from sqlmodel import Session
from app.database import get_session
from app.core.uow import UnitOfWork
from app.core.dependencies import require_role
from app.schemas.ingrediente import IngredienteCreate, IngredienteRead, IngredienteUpdate
from app.services.ingrediente_service import IngredienteService

router = APIRouter(prefix="/api/v1/ingredientes", tags=["Ingredientes"])


def get_service(session: Session = Depends(get_session)) -> IngredienteService:
    uow = UnitOfWork(session)
    return IngredienteService(uow)


@router.get("", response_model=List[IngredienteRead])
def listar_ingredientes(
    q: Optional[str] = Query(None, description="Buscar por nombre"),
    es_alergeno: Optional[bool] = Query(None, description="Filtrar por alérgeno"),
    service: IngredienteService = Depends(get_service),
):
    return service.get_all(q=q, es_alergeno=es_alergeno)


@router.get("/{id}", response_model=IngredienteRead)
def obtener_ingrediente(id: int, service: IngredienteService = Depends(get_service)):
    return service.get_by_id(id)


@router.post("", response_model=IngredienteRead, status_code=status.HTTP_201_CREATED)
def crear_ingrediente(
    data: IngredienteCreate,
    service: IngredienteService = Depends(get_service),
    _=Depends(require_role(["ADMIN"])),
):
    return service.create(data)


@router.patch("/{id}", response_model=IngredienteRead)
def actualizar_ingrediente(
    id: int,
    data: IngredienteUpdate,
    service: IngredienteService = Depends(get_service),
    _=Depends(require_role(["ADMIN"])),
):
    return service.update(id, data)


@router.delete("/{id}", status_code=status.HTTP_204_NO_CONTENT)
def eliminar_ingrediente(id: int, service: IngredienteService = Depends(get_service),
    _=Depends(require_role(["ADMIN"])),):
    service.delete(id)
